### ISSUE TITLE
Resource Lists - Mismatch between expected and actual JSON

### ISSUE THREAD

**Issue by bdchauvette** [phalt/pokeapi#94]
The [docs for resource lists](http://pokeapi.co/docsv2/#resource-lists) say that 'calling any api endpoint without a resource id or name will return a paginated list of available resources for that api' with a shape like:

```json
{
    "count": 365,
    "next": "http://pokeapi.co/api/v2/evolution-chain/?limit=20&offset=20",
    "previous": null,
    "results": [{
        "url": "http://pokeapi.co/api/v2/evolution-chain/1/"
    }]
}
```

When I hit a resource list endpoint, though, the results from the actual endpoints are different than expected. They currently seem to:

- return only the contents of what would be in the `results` object (i.e. an array of `{ url, ?name }` objects)
- ignore query parameters, so `limit` and `offset` seem to have no effect on what gets returned

For example, using `httpie`:

```bash
http http://pokeapi/api/v2/pokemon/
```
I get the following output:

```
HTTP/1.1 200 OK
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Type: application/json
Date: Mon, 08 Feb 2016 23:46:36 GMT
Server: nginx/1.1.19
Transfer-Encoding: chunked
Vary: Cookie
X-Frame-Options: SAMEORIGIN

[
    {
        "name": "bulbasaur",
        "url": "http://pokeapi.co/api/v2/pokemon/1/"
    },
    {
        "name": "ivysaur",
        "url": "http://pokeapi.co/api/v2/pokemon/2/"
    },
    {
        "name": "venusaur",
        "url": "http://pokeapi.co/api/v2/pokemon/3/"
    },

    [ . . . ]

    {
        "name": "latias-mega",
        "url": "http://pokeapi.co/api/v2/pokemon/10062/"
    },
    {
        "name": "latios-mega",
        "url": "http://pokeapi.co/api/v2/pokemon/10063/"
    }
]
```

I've curl'd, httpie'd, postman'd, and (far)fetch'd on a handful of different machines in different areas, so I don't *think* it's a problem on my end, but it's totally possible that I'm screwing up something completely simple and obvious :sweat_smile:

---

**tl;dr -- Are the docs for resource list endpoints outdated? Or are the docs correct and the API results are wrong? Or am I screwing up somewhere?**

Thanks! :grinning:

**Comment by zaneadix** [comment]
@bdchauvette Rest assured, you are sane and doing things correctly. We actually noticed this when we launched. We're really not sure what's causing the issue. Things work as expected when running locally in dev but live it's just spitting a list. Thank you for digging so far into it and formally logging a bug. Hopefully we can get this worked out soon. Is the issue impeding you in any way?

**Comment by bdchauvette** [comment]
No worries! I'm loving this API :grinning: 

I'm working on a JS wrapper, and wanted to create a mock API for offline testing. I came across this problem when working on mocking the resource lists, and I wasn't sure if I should try mocking the expected results, or the results that I was actually seeing.

**Comment by phalt** [comment]
@bdchauvette Great spot, and as @zaneadix said this is something that worked then suddenly stopped working (we should write a test to check in the future). It's likely just a configuration problem with the framework we use for the API layer (Django REST Framework).

You're welcome to investigate yourself if you want, but you've done quite a good job investigating this already. Myself or Zane can have a look and ship a fix soon :)

**Comment by bdchauvette** [comment]
I'll poke around the source a bit, but I'm not familiar with Django at all so I don't know how helpful my eyeballs will be :P

If I come across anything before you guys find a fix, I'll be sure to update this issue and/or submit a pull request.

**Comment by zaneadix** [comment]
Found this in the DRF docs.
Seems we just need to configure a page size.

<img width="595" alt="screen shot 2016-02-09 at 10 48 27 pm" src="https://cloud.githubusercontent.com/assets/7071557/12938292/b376e8de-cf7f-11e5-8f84-554ee36e8900.png">

If it doesn't get tackled sooner I'll roll it into my next pull request.

**Comment by zaneadix** [comment]
So it appears that PAGINATE_BY is an older setting and that PAGE_SIZE (which we're currently using) was introduced around 3.0. Is it possible that the version of DRF on the server is behind?

**Comment by phalt** [comment]
@zaneadix Running 3.0.0 at the moment. We need to change it to `PAGINATE_BY` in the settings.

@bdchauvette - wanna make your first pull request? :smile:

**Comment by bdchauvette** [comment]
@zaneadix's -- Good catch on the `PAGINATE_BY` / `PAGE_SIZE` issue!

It looks like `PAGINATE_BY` was [officially deprecated in 3.2](http://www.django-rest-framework.org/topics/3.2-announcement/#deprecations) and [removed in 3.3](http://www.django-rest-framework.org/topics/3.3-announcement/#deprecations). `PAGE_SIZE` & friends were [added in 3.1](http://www.django-rest-framework.org/topics/3.1-announcement/#pagination), so if the server is on 3.0, this would definitely explain the problem.

I'm not sure how feasible it would be to update the server's DRF version, but if you guys can't, I played around with two ways of handling this:

1. Put an upper limit on the `djangorestframework` version (e.g. 9143bb348ecbd45e152f8b32d84190897b8c5145)

2. Put a version check in `config/settings.py`, and set the REST_FRAMEWORK accordingly. (e.g. 88cecada24c99241e6b6376043991849b3f6da43)

Both of these seem to work, at least on my computer.

I personally think limiting the version number in `requirements.txt` is probably the cleaner option. The version check does things a bit more flexible and forward-compatible, but (1) it's pretty ugly, and (2) it opens the door for greater divergence later on between the dev and production environments.

@phalt & @zaneadix - let me know what you guys think of these options. If one of them seems decent, I'll put together a proper pull request for it :+1:

**Comment by phalt** [comment]
I prefer changing the settings to change a package requirement - we might need to upgrade DRF in the future.

**Comment by dalehurwitz** [comment]
@phalt @zaneadix it appears `?limit=` doesn't work but I'm assuming it's related to this issue?

**Comment by dalehurwitz** [comment]
Hey guys,

Just wondering if there's a plan for a fix anytime soon?

No pressure, just starting a project in the next few days and wanted to know if I should account for the bug.

Thanks :)

**Comment by zaneadix** [comment]
@dalehurwitz Hopefully we can get something up soon. we've already identified the issue. It really just depends on whether or not we can find time to push code up within the week.
